src/printer.js

```diff
@@ -2019,71 +2022,170 @@ function printMemberChain(node, options, print) {
   }
 }
 
+function isEmptyJSXElement(node) {
+  if (node.children.length === 0) return true;
+  if (node.children.length > 1) return false;
+
+  // if there is one child but it's just a newline, treat as empty
+  const value = node.children[0].value;
+  if (!/\S/.test(value) && /\n/.test(value)) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+function printJSXChildren(path, options, print) {
+  const n = path.getValue();
+  const children = [];
+
+  // using `map` instead of `each` because it provides `i`
+  path.map(function(childPath, i) {
+    const child = childPath.getValue();
+    const isLiteral = namedTypes.Literal.check(child);
+
+    if (isLiteral && typeof child.value === "string") {
+      if (/\S/.test(child.value)) {
+        const beginBreak = child.value.match(/^\s*\n/);
+        const endBreak = child.value.match(/\n\s*$/);
+        const beginSpace = child.value.match(/^\s+/);
+        const endSpace = child.value.match(/\s+$/);
+
+        if (beginBreak) {
+          children.push(hardline);
+        } else if (beginSpace) {
+          children.push(jsxWhitespace);
+        }
+
+        children.push(child.value.replace(/^\s+|\s+$/g, ""));
+
+        if (endBreak) {
+          children.push(hardline);
+        } else {
+          if (endSpace) children.push(jsxWhitespace);
+          children.push(softline);
+        }
+      } else if (/\n/.test(child.value)) {
+        // TODO: add another hardline if >1 newline appeared. (also above)
+        children.push(hardline);
+      } else if (/\s/.test(child.value)) {
+        // whitespace-only without newlines,
+        // eg; a single space separating two elements
+        children.push(jsxWhitespace);
+        children.push(softline);
+      }
+    } else {
+      children.push(print(childPath));
+
+      // add a line unless it's followed by a JSX newline
+      let next = n.children[i + 1];
+      if (!(next && /^\s*\n/.test(next.value))) {
+        children.push(softline);
+      }
+    }
+  }, "children");
+
+  return children;
+}
+
 function printJSXElement(path, options, print) {
   const n = path.getValue();
-  var openingLines = path.call(print, "openingElement");
 
-  let elem;
+  // turn <div></div> into <div />
```

Comment:
Are we sure we want to do this? It could be annoying if someone is in the middle of writing code and saves the file which autoformats the file. To add content they need to break it up again.